lib/cnf_graph.ml

```diff
@@ -305,3 +305,106 @@ let get_cnf g1 g2 max_time nbr_colors =
       assert (l1 = l2 && w1 = w2);
       []
 (*never return this list*)
+
+let pp_graph fmt g w l =
+  let rec aux_print_graph x y =
+    if y = l then ()
+    else if x = w then (
+      Format.fprintf fmt "\n";
+      aux_print_graph 0 (y + 1))
+    else (
+      Format.fprintf fmt "%d " g.(y).(x);
+      aux_print_graph (x + 1) y)
+  in
+  aux_print_graph 0 0
+
+let shuffle l =
+  let rec aux_shuffle = function
+    | [] -> []
+    | [ e ] -> [ e ]
+    | l ->
+        let left, right = List.partition (fun _ -> Random.bool ()) l in
+        List.rev_append (aux_shuffle left) (aux_shuffle right)
+  in
+  aux_shuffle l
+
+let get_all_cases w l =
+  let rec aux_get_all_cases x y acc =
+    if y = l then acc
+    else if x = w then aux_get_all_cases 0 (y + 1) acc
+    else aux_get_all_cases (x + 1) y ((x, y) :: acc)
+  in
+  aux_get_all_cases 0 0 []
+
+let my_mod x y =
+  let result = x mod y in
+  if result < 0 then result + y else result
+
+let get_possible_colors g w l x y =
+  let possible_colors = [ 1; 2; 3; 4 ] in
+  let rec aux_get_possible_colors colors acc =
+    match colors with
+    | [] -> acc
+    | n :: tl ->
+        if
+          g.(my_mod (y + 1) l).(x) = n
+          || g.(y).(my_mod (x + 1) w) = n
+          || g.(my_mod (y - 1) l).(x) = n
+          || g.(y).(my_mod (x - 1) w) = n
+        then aux_get_possible_colors tl acc
+        else aux_get_possible_colors tl (n :: acc)
+  in
+  aux_get_possible_colors possible_colors []
+
+let find_upper_recoloration g w l x y =
+  g.(my_mod (y - 1) l).(x) <- 0;
+  g.(my_mod (y - 1) l).(my_mod (x + 1) w) <- 0;
+  g.(y).(my_mod (x + 1) w) <- 0;
+  let rec try_to_color lst =
+    match lst with
+    | [] -> true
+    | (a, b) :: tl ->
+        let possible_colors = get_possible_colors g w l a b in
+        let rec try_colors colors =
+          match colors with
+          | [] -> false
+          | c :: tl_color ->
+              g.(b).(a) <- c;
+              if try_to_color tl then true else try_colors tl_color
+        in
+        try_colors possible_colors
+  in
+  if
+    not
+      (try_to_color
+         [
+           (x, my_mod (y - 1) l);
+           (my_mod (x + 1) w, my_mod (y - 1) l);
+           (x, y);
+           (my_mod (x + 1) w, y);
+         ])
+  then (
+    Format.printf "Error (%d, %d)\n" x y;
+    pp_graph Format.std_formatter g w l;
+    assert false)
+  else ()
+
+let fill_random_case_graph g w l x y =
+  let possible_colors = get_possible_colors g w l x y in
+  if possible_colors <> [] then g.(y).(x) <- List.hd (shuffle possible_colors)
```

Comment:
Is this really more efficient than just generating a random number and retrieving the i-th item from the list?

Reply:
I thought the same thing, I had the shuffle function so I did it with that, I told myself that on lists of no more than 4 items it doesn't change much, and since I like my shuffle function I wanted to use it in fact, if you absolutely want I can modify

Reply:
yes I hadn’t thought about the fact that there were max 4 elements, it’s not very important